### DISCUSSION TITLE
Issues with the ValidationContext when signing a PDF with PAdES-B

### DISCUSSION THREAD

**Discussion by jorgesisco** [MatthiasValvekens/pyHanko#252] [Q&A] [answered]
Hello, I have the following code:

```python
def PAdES(self, pdf_b64):
    input_pdf_data = base64.b64decode(pdf_b64)
    signer = signers.SimpleSigner.load_pkcs12(pfx_file=PFX_CERT, passphrase=PFX_CERT_PASS.encode('utf-8'))

    #Settign PAsES-B
    # Use BytesIO to treat the decoded PDF data as a file-like object
    with io.BytesIO(input_pdf_data) as doc, io.BytesIO() as out:
        w = IncrementalPdfFileWriter(doc)

        signature_meta = signers.PdfSignatureMetadata(
                field_name='Signature', md_algorithm='sha256',
                subfilter=SigSeedSubFilter.PADES,
                validation_context= ValidationContext(allow_fetching=True) #When I set it to false It works
            )

        signers.sign_pdf(
                w, signature_meta=signature_meta, signer=signer,
                output=out
            )

        # Get the signed PDF data from the BytesIO object and encode it as base64
        signed_pdf_data = out.getvalue()
        signed_pdf_base64 = base64.b64encode(signed_pdf_data).decode("utf-8")

        return signed_pdf_base64
```

I get an error related to the `ValidationContext` when setted to `True`, If I leave it to `False` I am able to do the signature but when I check in the european comission [checker](https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/validation) I it shows it passed the check but with the following Qualification Details:

  - The private key does not reside in a QSCD at (best) signing time!
  - The private key does not reside in a QSCD at issuance time!
 
When I set the `allow_fetching=True` I get the following error:
```bash
.
.
.
pyhanko_certvalidator.errors.InsufficientRevinfoError: The path could not be validated because the 
intermediate certificate 1 revocation checks failed: OCSP response is too recent

The above exception was the direct cause of the following exception:
.
.
.
pyhanko.sign.general.SigningError: The signer's certificate could not be validated
```

I this QSCD response related to the way I set the signer did I missed something? Or is it an issue with the certificates?

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @jorgesisco,

Regarding the OCSP error you're seeing: that's almost certainly due to a regression in `pyhanko-certvalidator` that is patched in the latest version. I'll copy what I posted in another thread just now:

> That being said, I have a potential explanation/solution in this case: last week I found a regression in pyhanko-certvalidator in the sanity checks on incoming revocation data. The way that usually manifests is that pyHanko rejects otherwise valid OCSP responses, which could cause the behaviour you're seeing. The bug in question should be fixed in pyhanko-certvalidator 0.21.x and up. Can you try upgrading pyhanko-certvalidator to 0.22.0? You might get an error from pip about dependency mismatches, but it's mostly safe to ignore---the breaking change is a very minor one that doesn't really matter for CLI users.
>
> PyHanko 0.18.0 will be released very soon, and it includes a bump for pyhanko-certvalidator, so if you can wait a few more days, that should also do the trick.
>
> If you still see the same issue after bumping pyhanko-certvalidator, let me know.


The QSCD attestation validation failure that you're seeing is generally not something that pyHanko can resolve. I'll spare you the boring details, but QSCD stands for "qualified signature creation device" (an eIDAS concept), and given that you're signing with a key on disk, you're almost surely not using a QSCD. Actually, if your certificate would have QSCD attestation extensions and the key does _not_ in fact reside in such a device, I'd have some strong words for whoever issued that certificate to you... ;)

**Reply by jorgesisco** [reply]
Thanks a lot for your answer!

Not that you ask but this is for integrating some services from a goverment entity, they have no mentions about QSCD in their docs. I am using the auth api cert as if it was an user certificate for signing the document since I am in a test enviroment I can do that, maybe is I sign with an user certificate It might have QSCD attestation, and if not, maybe they don't require that in the signature.

Thanks again for the answer!

**Reply by MatthiasValvekens** [reply]
Right, if you don't need qualified signatures (there are multiple signature tiers in eIDAS), then you can of course ignore those validation rules :).
